include/ringbuffer_utils.hpp

```diff
@@ -152,6 +153,151 @@ class PerfRingBufferReader {
+    if (n2 == 0) {
+      return {};
+    }
+
+    // \fixme{nsavoire} Not sure if spinlock is the best option here
+    lock(reinterpret_cast<std::atomic_flag *>(_rb.spinlock));
+
+    // No need for atomic operation, since we hold the lock
+    uint64_t writer_pos = *_rb.writer_pos;
+
+    uint64_t new_writer_pos = writer_pos + n2;
+    // Check that there is enough free space
+    if (_rb.mask < new_writer_pos - _tail) {
+      unlock(reinterpret_cast<std::atomic_flag *>(_rb.spinlock));
+      return {};
+    }
+
+    uint64_t head_linear = writer_pos & _rb.mask;
+    MPSCRingBufferHeader *hdr =
+        reinterpret_cast<MPSCRingBufferHeader *>(_rb.data + head_linear);
+
+    // Mark the sample as busy
+    hdr->size = n | MPSCRingBufferHeader::k_busy_bit;
+
+    // Atomic operation required to synchronize with reader load_acquire
+    __atomic_store_n(_rb.writer_pos, new_writer_pos, __ATOMIC_RELEASE);
+    unlock(reinterpret_cast<std::atomic_flag *>(_rb.spinlock));
+
+    return {reinterpret_cast<std::byte *>(hdr) + sizeof(MPSCRingBufferHeader),
+            n};
+  }
+
+  // Return true if notification to consumer is necesssary
+  // Notification is necessary only if consumer has caught up with producer
+  // (meaning tail afer commit is at or after head before commit)
+  bool commit(Buffer buf, bool discard = false) {
+    MPSCRingBufferHeader *hdr = reinterpret_cast<MPSCRingBufferHeader *>(
+        buf.data() - sizeof(MPSCRingBufferHeader));
+
+    // Clear busy bit
+    uint64_t new_size = hdr->size ^ MPSCRingBufferHeader::k_busy_bit;
+    if (discard) {
+      new_size |= MPSCRingBufferHeader::k_discard_bit;
+    }
+
+    __atomic_store_n(&hdr->size, new_size, __ATOMIC_RELAXED);
+
+    _tail = __atomic_load_n(_rb.reader_pos, __ATOMIC_ACQUIRE);
+    uint64_t tail_linear = _tail & _rb.mask;
+    return tail_linear ==
+        static_cast<uint64_t>(reinterpret_cast<std::byte *>(hdr) - _rb.data);
+  }
+
+private:
+  RingBuffer &_rb;
+  uint64_t _tail;
+};
+
+class MPSCRingBufferReader {
+public:
+  explicit MPSCRingBufferReader(RingBuffer &rb) : _rb(rb) {
+    assert(rb.type == RingBufferType::kMPSCRingBuffer);
+    _head = __atomic_load_n(rb.writer_pos, __ATOMIC_ACQUIRE);
+    _tail = *_rb.reader_pos;
+    _initial_tail = _tail;
+    assert(_tail <= _head);
+  }
+
+  ~MPSCRingBufferReader() {
+    if (_tail != _initial_tail) {
+      __atomic_store_n(_rb.reader_pos, _tail, __ATOMIC_RELEASE);
+    }
+  }
+
+  inline size_t available_size() const { return _head - _tail; }
+
+  ConstBuffer read_sample() {
+    size_t n = _head - _tail;
+    if (n == 0) {
+      return {};
+    }
+
+    uint64_t tail_linear = _tail & _rb.mask;
+    std::byte *start = _rb.data + tail_linear;
+    MPSCRingBufferHeader *hdr = reinterpret_cast<MPSCRingBufferHeader *>(start);
+    uint64_t sz = __atomic_load_n(&hdr->size, __ATOMIC_ACQUIRE);
+
+    // Sample not committed yet, bail out
+    if (MPSCRingBufferHeader::is_busy(sz)) {
```

Comment:
Makes sense to me.  Leave it to the calling application to determine how to handle the scenario where the ringbuffer is in the middle of a reserve-commit sequence.

Reply:
For the calling application, we behave as if the ringbuffer was empty, we don't convey the information that a sample is being finalized.
We could try reading the next samples but this would add a lot more complexity. WDYT ?